### PR TITLE
Feature/network

### PR DESCRIPTION
**Pull request by minsulee2**
Updates:
- Separate p2p server and p2p client
- Refactor p2p server
- Refactor tracker server
- Fix some minor bugs

TODOs:
- Separate server and client more precisely
- Separate execute and broadcast functions
- Do not send msg from consensus
- Heartbeat
- sign and verify

### CODE REVIEW

File: client/index.js [hunk 1]
```diff
@@ -558,3 +566,22 @@ function validateVersion(req, res, next) {
     next();
   }
 }
+
+function matchConnections(numConnection, numOutbound, numInbound) {
+  let maxConnection = numConnection;
+  let maxOutbound = numOutbound;
+  let maxInbound = numInbound;
+  if (0 > numConnection || numConnection < 5) {
```
Comment (platfowner) [inline client/index.js:574]: Math.max(numConnection, 5)?

File: client/index.js [hunk 2]
```diff
@@ -558,3 +566,22 @@ function validateVersion(req, res, next) {
...
+    maxConnection = 5;
+  }
+  if (0 > numOutbound || numOutbound < 2) {
```
Comment (platfowner) [inline client/index.js:577]: Math.max(numOutbound, 2)?

File: client/index.js [hunk 3]
```diff
@@ -558,3 +566,22 @@ function validateVersion(req, res, next) {
...
+    maxOutbound = 2;
+  }
+  if (maxConnection < maxOutbound) {
```
Comment (platfowner) [inline client/index.js:580]: Math.min(maxOutbound, maxConnection)?

File: client/index.js [hunk 4]
```diff
@@ -558,3 +566,22 @@ function validateVersion(req, res, next) {
...
+    maxOutbound = maxConnection;
+  }
+  if (0 > numInbound || numInbound < numConnection - numOutbound) {
```
Comment (platfowner) [inline client/index.js:583]: `const maxInbound = Math.max(numInbound, numConnection - numOutbound)`?

File: common/constants.js [hunk 5]
```diff
@@ -53,6 +53,8 @@ const MessageTypes = {
   CHAIN_SEGMENT: 'CHAIN_SEGMENT',
   CHAIN_SEGMENT_REQUEST: 'CHAIN_SEGMENT_REQUEST',
   CONSENSUS: 'CONSENSUS',
+  ACCOUNT: 'ACCOUNT',
```
Comment (platfowner) [inline common/constants.js:56]: minor: what about adding '_RESPONSE' suffix to 'ACCOUNT' and 'CHAIN_SEGMENT' for readability?

File: consensus/constants.js [hunk 6]
```diff
@@ -3,11 +3,15 @@ const ConsensusConsts = {
   DAY_MS: 86400000,
   EPOCH_MS: 3000,
   MAX_CONSENSUS_STATE_DB: 1000,
+  HEALTH_THRESHOLD_EPOCH: 600, // 600 epochs = 10 minutes
   GENESIS_STAKE: 250,
   MIN_STAKE_PER_VALIDATOR: 100,
   // TODO(lia): Use a separate genesis json file for consensus params and deprecate NUM_VALIDATORS
   MIN_NUM_VALIDATORS: process.env.NUM_VALIDATORS ? Number(process.env.NUM_VALIDATORS) : 3,
-  HEALTH_THRESHOLD_EPOCH: 600 // 600 epochs = 10 minutes
+  HEALTH_THRESHOLD_EPOCH: 600, // 600 epochs = 10 minutes
```
Comment (platfowner) [inline consensus/constants.js:11]: dup definition?

File: consensus/constants.js [hunk 7]
```diff
@@ -3,11 +3,15 @@ const ConsensusConsts = {
...
+  INITIAL_MAX_CONNECTION: process.env.MAX_CONNECTION ? Number(process.env.MAX_CONNECTION) : 5,
```
Comment (platfowner) [inline consensus/constants.js:12]: Unless these values are related to consensus, what about moving out of this enum to somewhere in common/constants.h?

File: client/index.js [hunk 8]
```diff
@@ -558,3 +566,22 @@ function validateVersion(req, res, next) {
     next();
   }
 }
+
+function matchConnections(numConnection, numOutbound, numInbound) {
```
Comment (platfowner) [inline client/index.js:570]: What about moving this under server/index.js as a static function?

File: server/index.js [hunk 9]
```diff
@@ -792,6 +583,392 @@ class P2pServer {
     await sendTxAndWaitForFinalization(parentChainEndpoint, shardInitTx, ownerPrivateKey);
     logger.info(`setUpDbForSharding success`);
   }
+}
+
+class P2pClient {
```
Comment (platfowner) [inline server/index.js:588]: What about moving this into a separate file?

File: server/index.js [hunk 10]
```diff
@@ -792,6 +583,392 @@ class P2pServer {
...
+  }
+
+  buildManagedPeersInfo() {
+    return {
+      outbound: this.outbound,
+      inbound: this.server.inbound
+    };
+  }
+
+  // TODO(seo): Add sharding status.
```
Comment (platfowner) [inline server/index.js:626]: Could you remove this for me? It's already done.

File: tracker-server/index.js [hunk 11]
```diff
@@ -109,221 +77,131 @@ const server = new WebSocketServer({
...
   }
+}
 
-  addPeer(peer) {
-    if (peer && peer.address !== this.address && !this.managedPeers[peer.address] &&
-        !this.unmanagedPeers[peer.address]) {
-      this.managedPeers[peer.address] = true;
-      peer.unmanagedPeers[this.address] = true;
+function getPeerCandidates(myself, candidates) {
+  Object.keys(PEER_NODES).forEach(address => {
```
Comment (platfowner) [inline tracker-server/index.js:156]: minor: `for (const address in PEER_NODES) {`?

File: tracker-server/index.js [hunk 12]
```diff
@@ -109,221 +77,131 @@ const server = new WebSocketServer({
...
-      logger.info(`>> Message to node [${abbrAddr(node.address)}]: ` +
-          `${JSON.stringify(msgToNode, null, 2)}`)
-      ws.send(JSON.stringify(msgToNode));
-      printNodesInfo();
-    } catch (error) {
-      logger.error(error.stack);
+    const nodeInfo = JSON.parse(message);
+    WS_LIST[ws.uuid] = nodeInfo.address;
+    if (PEER_NODES[nodeInfo.address]) {
+      PEER_NODES[nodeInfo.address] = nodeInfo;
```
Comment (platfowner) [inline tracker-server/index.js:92]: can be pulled out of if-else block?

Comment (platfowner) [reply tracker-server/index.js:92]: ah, location is set. please ignore the comment above.

File: consensus/index.js [hunk 13]
```diff
@@ -216,12 +216,13 @@ class Consensus {
         // prevent the node from getting/handling messages properly.
         // this.node.status = BlockchainNodeStatus.SYNCING;
 
-        this.server.requestChainSegment(this.node.bc.lastBlock());
+        // TODO(minsu): FIXME!!
```
Comment (liayoo) [inline consensus/index.js:219]: Could you be a bit more specific?

File: tracker-server/index.js [hunk 14]
```diff
@@ -4,89 +4,57 @@ const geoip = require('geoip-lite');
 const express = require('express');
 const jayson = require('jayson');
 const _ = require('lodash');
+const { v4: uuidv4 } = require('uuid');
 const logger = require('../logger')('TRACKER_SERVER');
 
 const P2P_PORT = process.env.P2P_PORT || 5000;
 const PORT = process.env.PORT || 8080;
-const MAX_NUM_PEERS = 2;
 const PEER_NODES = {};
```
Comment (liayoo) [inline tracker-server/index.js:12]: suggestion) How about using camelCase? Currently, I think it implies that it's 'constant'. Same goes for WS_LIST.

File: server/index.js [hunk 15]
```diff
@@ -556,87 +399,35 @@ class P2pServer {
...
-  broadcastConsensusMessage(msg) {
-    logger.debug(`SENDING: ${JSON.stringify(msg)}`);
-    this.sockets.forEach((socket) => {
-      socket.send(JSON.stringify({
-        type: MessageTypes.CONSENSUS,
-        message: msg,
-        protoVer: CURRENT_PROTOCOL_VERSION
-      }));
-    });
+  removeFromListIfExists(address) {
```
Comment (liayoo) [inline server/index.js:423]: suggestion) removeFromListIfExists -> removeFromInboundListIfExists, removeFromOutboundListIfExists

File: server/index.js [hunk 16]
```diff
@@ -52,22 +51,21 @@ const HEARTBEAT_INTERVAL_MS = 1000;
...
-// TODO(seo): Sign messages to tracker or peer.
+// TODO(minsu): Sign messages to tracker or peer.
 class P2pServer {
-  constructor(node, minProtocolVersion, maxProtocolVersion) {
-    this.internalIpAddress = null;
-    this.externalIpAddress = null;
-    this.trackerWebSocket = null;
+  constructor (node, minProtocolVersion, maxProtocolVersion, maxConnection, maxOutbound, maxInbound)
+  {
     this.server = null;
```
Comment (platfowner) [inline server/index.js:58]: As this is a web socket server, what about giving more specific name like this.wsServer? Just this.server sounds like it's a counterpart of this.client?

File: client/index.js (hunk 8 above)
Comment (platfowner) [reply client/index.js:570]: And if the variables from this function are not used except initializing p2pServer, would it be better to hide this inside the p2pServer logic?

**Comment by platfowner** [comment]
Thanks for the hard work, I have several basic-level questions:
- I see that sendChainSegment() is a part of the client logic.
  What I assumed was that server handles requests from other peers while client sends requests to peers (also handling the responses from the servers).
  If it's correct, should this function a part of the server class? 
  
  Maybe it's caused by trying to separate code for the client and the server. Some functions (e.g. broadcastTransaction()) may lay in an overlapping area? 
  
- For the separation of the server part and the client code, I'm not quite sure about its benefits. It seems that we see clear benefits from web socket channel separation for the client and the server, what would be the benefits of having separate code (i.e., class) for the client and the server?

File: server/index.js (hunk 9 above)
Comment (platfowner) [reply server/index.js:588]: Until the refactoring is on a finalization stage, I think it's just fine to have a todo here.

**Comment by platfowner** [comment]
Discussed offline with the following agreements:
- The server handles requests (CHAIN_SEGMENT_REQUEST, ACCOUNT_REQUEST), while the client sends requests and broad cast messages (TRANSACTION, CONSENSUS)
- Let have the server logic in P2PServer and the client logic in P2PClient
- For further refactoring, let's discuss in a meeting with @liayoo

File: client/index.js (hunk 1 above)
Comment (minsulee2) [reply client/index.js]: Done!

File: client/index.js (hunk 2 above)
Comment (minsulee2) [reply client/index.js]: Done!

File: client/index.js (hunk 3 above)
Comment (minsulee2) [reply client/index.js]: Done!

File: client/index.js (hunk 4 above)
Comment (minsulee2) [reply client/index.js]: Done!

File: common/constants.js (hunk 5 above)
Comment (minsulee2) [reply common/constants.js]: Yeah, this would be better. applied em!

File: consensus/constants.js (hunk 6 above)
Comment (minsulee2) [reply consensus/constants.js]: That was , change.

File: consensus/constants.js (hunk 7 above)
Comment (minsulee2) [reply consensus/constants.js]: It moved thanks!

File: client/index.js (hunk 8 above)
Comment (minsulee2) [reply client/index.js]: It moved to p2p server! thanks!

File: server/index.js (hunk 9 above)
Comment (minsulee2) [reply server/index.js]: Separted!

File: server/index.js (hunk 10 above)
Comment (minsulee2) [reply server/index.js]: Done!

File: tracker-server/index.js (hunk 11 above)
Comment (minsulee2) [reply tracker-server/index.js]: Done!

File: consensus/index.js (hunk 13 above)
Comment (minsulee2) [reply consensus/index.js]: Added some comments on it!

File: tracker-server/index.js (hunk 14 above)
Comment (minsulee2) [reply tracker-server/index.js]: Updated!

File: server/index.js (hunk 15 above)
Comment (minsulee2) [reply server/index.js]: Updated!

File: server/index.js (hunk 16 above)
Comment (minsulee2) [reply server/index.js]: Sounds gooooood. Updated!! it is now way clearer!

**Comment by minsulee2** [comment]
As discussed off-line, the further refactoring especially connecting to consensus parts will be discussed together and fixed in a next PR.
PTAL @platfowner @liayoo

File: consensus/constants.js [hunk 17]
```diff
@@ -3,6 +3,7 @@ const ConsensusConsts = {
   DAY_MS: 86400000,
   EPOCH_MS: 3000,
   MAX_CONSENSUS_STATE_DB: 1000,
+  HEALTH_THRESHOLD_EPOCH: 600, // 600 epochs = 10 minutes
```
Comment (platfowner) [inline consensus/constants.js:6]: Is this used somewhere?

Comment (platfowner) [reply consensus/constants.js:6]: or dup definition?

File: server/client.js [hunk 18]
```diff
@@ -0,0 +1,427 @@
...
+      this.setTrackerEventHandlers();
+      this.setIntervalForTrackerUpdate();
+    });
+    this.trackerWebSocket.on('error', (error) => {
+      logger.error(`Error in communication with tracker (${TRACKER_WS_ADDR}): ` +
+        `${JSON.stringify(error, null, 2)}`);
+    });
+  }
+
+  sendChainSegment(socket, chainSegment, number, catchUpInfo) {
```
Comment (platfowner) [inline server/client.js:142]: As this is called only from the server-side, why don't you move this under P2PServer class?

File: server/index.js [hunk 19]
```diff
@@ -793,20 +586,14 @@ class P2pServer {
...
-  }
-
-  clearIntervalHeartbeat(address) {
-    clearInterval(this.managedPeersInfo[address].intervalHeartbeat);
-    this.managedPeersInfo[address].intervalHeartbeat = null;
+  // NOTE(minsu): the total number of connection is up to more than 5 without limit.
+  // maxOutbound is for now limited equal or less than 2.
+  // maxInbound is a rest of connection after maxOutbound is set.
+  static matchConnections(numConnection, numOutbound, numInbound) {
+    const maxConnection = Math.max(numConnection, 5);
```
Comment (platfowner) [inline server/index.js:593]: What about defining 5 and 2 as constants?

File: tracker-server/index.js [hunk 20]
```diff
@@ -109,221 +77,130 @@ const server = new WebSocketServer({
...
+    printNodesInfo();
   });
 
+  // TODO(minsu): code should be setup ex) code === 1006: SIGINT
   ws.on('close', (code) => {
-    logger.info(`\nDisconnected from node [${node ? abbrAddr(node.address) : 'unknown'}] ` +
+    const address = wsList[ws.uuid];
+    logger.info(`\nDisconnected from node [${address ? abbrAddr(address) : 'unknown'}] ` +
         `with code: ${code}`);
-    PEER_NODES[node.address].isLive = false;
```
Comment (platfowner) [inline tracker-server/index.js:149]: We used to be using PEER_NODES to track node status in a non-volatile way, i.e., even when some of them die /peer_nodes API still returns their last status (e.g. memory status, disk status, etc) for debugging purpose. In such circumstance, isLive flag was an indicator for each node's liveness. 
If we just reset the status data when some nodes are disconnected, their status will be lost.

File: consensus/constants.js (hunk 17 above)
Comment (minsulee2) [reply consensus/constants.js]: Oh, my apology! It was dup definition due to conflicts.

File: server/client.js (hunk 18 above)
Comment (minsulee2) [reply server/client.js]: Right! I was so confusing! Fixed!

File: server/index.js (hunk 19 above)
Comment (minsulee2) [reply server/index.js]: Done!

File: tracker-server/index.js (hunk 20 above)
Comment (minsulee2) [reply tracker-server/index.js:149]: I'd thought it was next up with heartbeat. But, it was necessary! I've revived the isAlive status and left TODO for isAlive as well.

**Comment by minsulee2** [comment]
@platfowner PTAL. All addressed!

File: client/index.js [hunk 21]
```diff
@@ -39,15 +40,22 @@ if (!semver.valid(CURRENT_PROTOCOL_VERSION)) {
...
+} = P2pServer.matchConnections(PeerConnections.INITIAL_MAX_CONNECTION,
+    PeerConnections.INITIAL_MAX_OUTBOUND, PeerConnections.INITIAL_MAX_INBOUND);
 
 const app = express();
 app.use(express.json()); // support json encoded bodies
 
 const node = new BlockchainNode();
-const p2pServer = new P2pServer(node, minProtocolVersion, maxProtocolVersion);
+const p2pServer = new P2pServer(node, minProtocolVersion, maxProtocolVersion,
+    maxConnection, maxOutbound, maxInbound);
```
Comment (platfowner) [inline client/index.js:58]: minor: Unless we have a plan to alter these values for some cases (e.g. testing) in near future, it seems that these values can just be set implicitly in the constructor of P2pServer?

File: common/constants.js [hunk 22]
```diff
@@ -351,7 +354,18 @@ const FeatureFlags = {
   enableStateVersionOpt: true,
   // Enables state version renaming.
   enableVersionRenaming: true,
-}
+};
+
+/**
+ * Connection info.
+ */
+const PeerConnections = {
```
Comment (platfowner) [inline common/constants.js:362]: what about defining only initial and default values here and handle the actual values outside of this enum definition? e.g.,

const PeerConnections = {
  DEFAULT_MAX_CONNECTION: 5,
  ...
  INITIAL_MAX_CONNECTION: 5,
  ...
}

  static matchConnections() {
    const numConnection = process.env.MAX_CONNECTION ? Number(process.env.MAX_CONNECTION) : PeerConnections.INITIAL_MAX_CONNECTION;
    ...
    const maxConnection = Math.max(numConnection, PeerConnections.DEFAULT_MAX_CONNECTION);
    ...
    return { maxConnection, maxOutbound, maxInbound };
  }

File: server/index.js [hunk 23]
```diff
@@ -793,20 +597,14 @@ class P2pServer {
...
-
-  clearIntervalHeartbeat(address) {
-    clearInterval(this.managedPeersInfo[address].intervalHeartbeat);
-    this.managedPeersInfo[address].intervalHeartbeat = null;
+  // NOTE(minsu): the total number of connection is up to more than 5 without limit.
+  // maxOutbound is for now limited equal or less than 2.
+  // maxInbound is a rest of connection after maxOutbound is set.
+  static matchConnections(numConnection, numOutbound, numInbound) {
+    const maxConnection = Math.max(numConnection, PeerConnections.DEFAULT_MAX_CONNECTION);
+    const maxOutbound = Math.min(numOutbound, PeerConnections.DEFAULT_MAX_OUTBOUND);
```
Comment (platfowner) [inline server/index.js:605]: It seems that DEFAULT_MAX_CONNECTION and DEFAULT_MAX_OUTBOUND values are used as some limit values. If it's correct, what about naming so? e.g. MAX_CONNECTION_LIMIT, MAX_OUTBOUND_LIMIT

File: server/index.js [hunk 24]
```diff
@@ -793,20 +597,14 @@ class P2pServer {
...
-        ws.ping();
-      });
-    }, HEARTBEAT_INTERVAL_MS);
-  }
-
-  clearIntervalHeartbeat(address) {
-    clearInterval(this.managedPeersInfo[address].intervalHeartbeat);
-    this.managedPeersInfo[address].intervalHeartbeat = null;
+  // NOTE(minsu): the total number of connection is up to more than 5 without limit.
+  // maxOutbound is for now limited equal or less than 2.
```
Comment (platfowner) [inline server/index.js:601]: It seems that the DEFAULT_MAX_OUTBOUND is an upper bound? If so, `Math.max(numOutbound, PeerConnections.DEFAULT_MAX_OUTBOUND)`?

File: tracker-server/index.js (hunk 20 above)
Comment (platfowner) [reply tracker-server/index.js:149]: Thanks!

File: common/constants.js (hunk 22 above)
Comment (minsulee2) [reply common/constants.js:362]: Updated!

File: server/index.js (hunk 23 above)
Comment (minsulee2) [reply server/index.js]: Updated!

File: server/index.js (hunk 24 above)
Comment (minsulee2) [reply server/index.js:601]: Updated!

File: client/index.js (hunk 21 above)
Comment (minsulee2) [reply client/index.js:53]: Oh, I'd missed this part before. I thought providing connection status is good idea thru REST api. What do you think?

**Comment by minsulee2** [comment]
PTAL! @platfowner I feel almost there!

File: client/index.js (hunk 21 above)
Comment (platfowner) [reply client/index.js:53]: Do you want to expose the return values of matchConnections() or the current P2PServer's member variables? e.g.
```
  constructor (node, minProtocolVersion, maxProtocolVersion, maxConnection, maxOutbound, maxInbound)
  {
    ...
    this.maxConnection = maxConnection;
    this.maxOutbound = maxOutbound;
    this.maxInbound = maxInbound;
``` 

(Maybe the values don't change after initial setting) 
If it's the latter case, a new member function like getConnectionStatus() can be added for the purpose, which returns the three values plus some connection related numbers?

Comment (minsulee2) [reply client/index.js]: Done!

Comment (minsulee2) [reply client/index.js]: PTAL :)

File: server/client.js [hunk 25]
```diff
@@ -0,0 +1,417 @@
...
+    }, RECONNECT_INTERVAL_MS);
+  }
+
+  clearIntervalForTrackerConnection() {
+    clearInterval(this.intervalConnection);
+    this.intervalConnection = null;
+  }
+
+  // TODO(minsu): maxConnection in client/index.js, maxOutbound in p2pClient, maxInbound in p2pServer
+  getConnectionInfo() {
```
Comment (platfowner) [inline server/client.js:44]: Can be replaced with server.getConnectionInfo()?

**Review by platfowner — APPROVED** [review]
LGTM with some minor comments.
Thanks!

File: server/client.js (hunk 25 above)
Comment (minsulee2) [reply server/client.js]: Done!

**Comment by minsulee2** [comment]
Thanks for getting through the long journey together! I am merging it now.

**Pull request merged**